pkg/operation/etcdencryption/encryptionkeys.go

```diff
@@ -0,0 +1,116 @@
+// Copyright (c) 2019 SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package encryptionconfiguration
+
+import (
+	"crypto/rand"
+	"encoding/base64"
+	"fmt"
+	"strconv"
+	"strings"
+	"time"
+
+	apiserverconfigv1 "k8s.io/apiserver/pkg/apis/config/v1"
+)
+
+func createEncryptionKey() (apiserverconfigv1.Key, error) {
+	var key apiserverconfigv1.Key
+	keyString, err := createKeySecret()
+	if err != nil {
+		return key, err
+	}
+	if len(keyString) == 0 {
+		return key, fmt.Errorf("generated random encryption key is of unexpected length 0")
+	}
+	key.Name = createKeyName()
+	key.Secret = keyString
+	return key, nil
+}
+
+func createKeyName() string {
+	nanos := time.Now().UnixNano()
+	return ecKeyPrefix + strconv.FormatInt(nanos, 10)
+}
+
+func createKeySecret() (string, error) {
+	keyArray := make([]byte, ecKeySecretLen)
+	_, err := rand.Read(keyArray)
+	if err != nil {
+		return "", fmt.Errorf("error while obtaining random data: %v", err)
+	}
+	sEnc := base64.StdEncoding.EncodeToString(keyArray)
+	return sEnc, nil
+}
+
+func parseTimestampFromKeyName(keyName string) (int64, error) {
+	nameConsistent, err := isKeyNameConsistent(keyName)
+	if err != nil {
+		return 0, err
+	} else if !nameConsistent {
+		return 0, nil
+	}
+	timestampstring := keyName[len(ecKeyPrefix):len(keyName)]
+	timestamp, _ := strconv.ParseInt(timestampstring, 10, 64)
+	return timestamp, nil
+}
+
+func isKeyConsistent(key apiserverconfigv1.Key) (bool, error) {
```

Comment:
Why `(bool, error)`? The only state you'll ever have is valid or invalid with reason -> `error` should suffice as return value.
Btw, I'd also recommend calling this `ValidateKey` instead of that and also moving it to proper validation code.

Reply:
changed signature to return solely error and renamed method as you suggested. Am not sure what you mean by moving it to "proper validation code" however, since I actually thought the validation of the item as being part of the item itself ...

Reply:
Take a look at `pgk/apis/garden/validation`: For each type, we provide validation functions that ensure that no invalid object can get into the api server. We defined these validations for all of our types and they also report multiple issues with a resource. If you have proper validations in place, at retrieval time of an object you'll never have an invalid one and you're in line with how other k8s objects work.